### PR TITLE
Add model training loss convergence benchmark

### PR DESCRIPTION
**Pull request by yanboliang**
This example follows how users write their own end to end model training/evaluation pipeline using _huggingface_. The loss function is in convergence w/o Dynamo:
```
Loss history : [1.638656200170517, 1.6167153787612916, 1.591713446378708, 1.5690117526054381, 1.5178297901153563, 1.4247627103328704, 1.325632695555687, 1.23151354432106, 1.2123524838685988, 1.1726598739624023, 1.0923227965831757, 1.1367704308032989, 1.0908222246170043, 1.1158152759075164, 1.0315754282474519, 1.06554756462574, 1.0479815247654916, 1.0996541720628739, 1.0450058096647263, 1.0318933933973313, 1.0339357393980027, 1.0286797326803208, 1.052253406047821, 1.0163359659910203, 0.9940995579957962, 0.988142160475254, 0.9932016897201538, 0.9409900230169296, 0.9778001809120178, 1.0251611429452896, 0.9572644853591918, 1.013562046289444, 0.9348490306735039, 0.9323300030827523, 1.0125887167453766, 0.9674993979930878]
```  
There are several bugs if we enable Dynamo, I'm working on fixing them now. Posting it before bug-free is to collect feedback on if this framework is what users usually does and then we can optimize Dynamo to provide the best performance and experience with least onboarding cost.

cc @anijain2305

### CODE REVIEW

**Comment by yanboliang** [comment]
Update: This training benchmark works well and ready for review now.

File: benchmarks/training_loss.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+
+def model_training_evaluation(
+    train_dataloader, eval_dataloader, model, optimizer, num_epochs
+):
+    model.to(device)
+    model.train()
+    loss_history = []
+    # Support backends: eager, aot_nop and aot_nvfuser
+    opt_training_iter_fn = torchdynamo.optimize("eager")(training_iter_fn)
```
Comment (anijain2305) [inline benchmarks/training_loss.py:65]: pass `backend` and support it in arg.parser?

File: benchmarks/training_loss.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+from datasets import load_dataset
+from datasets import load_metric
+from torch.optim import SGD
+from torch.utils.data import DataLoader
+from transformers import AutoModelForSequenceClassification
+from transformers import AutoTokenizer
+
+import torchdynamo
+
+torchdynamo.config.fake_tensor_propagation = False
```
Comment (anijain2305) [inline benchmarks/training_loss.py:13]: Can we turn this on? Or list down the issues that we need to fix to turn it on?

File: benchmarks/training_loss.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        predictions = torch.argmax(logits, dim=-1)
+        metric.add_batch(predictions=predictions, references=batch["labels"])
+
+    print("Loss history : " + str(loss_history))
+    print("Accuracy : " + str(metric.compute()))
+
+
+if __name__ == "__main__":
+    train_dataloader, eval_dataloader = data_processing(num_samples)
+    model = AutoModelForSequenceClassification.from_pretrained(
```
Comment (anijain2305) [inline benchmarks/training_loss.py:94]: This can go in next PR. But, I have a few suggestions

1) We should provide some glue code, so that we can say test PASSED or FAILED. Like, run native pytorch and optimize("some_backend") have close-enough loss after some iterations.
2) Ability to specify more models.
3) Ability to specify more optimizers.

**Review by anijain2305 — CHANGES_REQUESTED** [review]

File: benchmarks/training_loss.py (hunk 2 above)
Comment (voznesenskym) [reply benchmarks/training_loss.py:13]: Can we use @patch to set this, instead of setting it globally?

File: benchmarks/training_loss.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+from torch.optim import SGD
+from torch.utils.data import DataLoader
+from transformers import AutoModelForSequenceClassification
+from transformers import AutoTokenizer
+
+import torchdynamo
+
+torchdynamo.config.fake_tensor_propagation = False
+
+# You will download around 84G dataset if you run this end to end training/evaluation example.
```
Comment (voznesenskym) [inline benchmarks/training_loss.py:15]: Is there a way to cache this artifact in CI? Or maybe have a flag to run a smaller version?

**Comment by voznesenskym** [comment]
How useful is printing loss history? How do we use this benchmark to tell if dynamo changes convergence on its own? with a backend? As it is, I am not sure who would run this benchmark and get value from it...

My $0.02 (take it or leave it) Is we already have training benchmarks... maybe we can extend those to report loss / entropy / normalized cross-entropy (w/e you like) at a given epoch (say, 1000?) We can leverage the infra @anijain2305 made to make our awesome "Performance Dashboard" and just add a column / record for $some_loss_measure at $some_k_epochs. Thoughts?

**Comment by anijain2305** [comment]
> How useful is printing loss history? How do we use this benchmark to tell if dynamo changes convergence on its own? with a backend? As it is, I am not sure who would run this benchmark and get value from it...
> 
> My $0.02 (take it or leave it) Is we already have training benchmarks... maybe we can extend those to report loss / entropy / normalized cross-entropy (w/e you like) at a given epoch (say, 1000?) We can leverage the infra @anijain2305 made to make our awesome "Performance Dashboard" and just add a column / record for $some_loss_measure at $some_k_epochs. Thoughts?

We don't do multiple iteration training in the current benchmarking infra (though it would be ideal to put all this one place). We run just one iteration of fwd-bwd there. There are multiple issues behind it - (1) As it stands now, the goal of current benchmarking infra is to aid development, find bugs in single iter, and quickly iterate. (2) There is no easy way to get valid dataloader for all the benchmarks (or even a few) to train to completion.

This created a gap, where we were not testing Dynamo on end-to-end training. So, I suggested Yanbo to create this. This has exposed multiple bugs. The main goal here is to support training for few models + optimizers and convince ourselves that Dynamo+inductor actually work for long training jobs.

**Comment by voznesenskym** [comment]
> > How useful is printing loss history? How do we use this benchmark to tell if dynamo changes convergence on its own? with a backend? As it is, I am not sure who would run this benchmark and get value from it...
> 
> > 
> 
> > My $0.02 (take it or leave it) Is we already have training benchmarks... maybe we can extend those to report loss / entropy / normalized cross-entropy (w/e you like) at a given epoch (say, 1000?) We can leverage the infra @anijain2305 made to make our awesome "Performance Dashboard" and just add a column / record for $some_loss_measure at $some_k_epochs. Thoughts?
> 
> 
> 
> We don't do multiple iteration training in the current benchmarking infra (though it would be ideal to put all this one place). We run just one iteration of fwd-bwd there. There are multiple issues behind it - (1) As it stands now, the goal of current benchmarking infra is to aid development, find bugs in single iter, and quickly iterate. (2) There is no easy way to get valid dataloader for all the benchmarks (or even a few) to train to completion.
> 
> 
> 
> This created a gap, where we were not testing Dynamo on end-to-end training. So, I suggested Yanbo to create this. This has exposed multiple bugs. The main goal here is to support training for few models + optimizers and convince ourselves that Dynamo+inductor actually work for long training jobs. 
> 
> 
> 
> 

Makes sense! We should maybe add a followup to track convergence across backends?

**Comment by anijain2305** [comment]
> We should maybe add a followup to track convergence across backends?

Yes, I agree with that. We should refactor to say PASS or FAIL and ability to try different backends. Otherwise, its too much eyeballing for now. Hopefully, we can use this script in the future to publish numbers like the way Nvidia does it - https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch/LanguageModeling/BERT#results

File: benchmarks/training_loss.py (hunk 2 above)
Comment (yanboliang) [reply benchmarks/training_loss.py]: > Can we turn this on? Or list down the issues that we need to fix to turn it on?

This issue has been tracked by https://github.com/pytorch/torchdynamo/issues/1076

> Can we use @patch to set this, instead of setting it globally?

I think this is both for benchmark and example, we use this to show how to enable TorchDynamo end to end. So would prefer not to add decorator on fn in case confusing users. And this is a temporary config, will remove it when we fix the above issue.

File: benchmarks/training_loss.py (hunk 3 above)
Comment (yanboliang) [reply benchmarks/training_loss.py:135]: Sounds good! I added the ability to specify backend/optimizer and several model training parameters. We can add the ability to compare training loss for native pytorch vs TorchDynamo in follow-ups.

File: benchmarks/training_loss.py (hunk 4 above)
Comment (yanboliang) [reply benchmarks/training_loss.py:16]: I was thinking to cache it in CI, but it's too big IMO. I have two concerns to have a smaller version:
* We need to generate the smaller version by ourself, so it's not a standard HF training pipeline any more. This is also used as an example/tutorial to help users onboard their models. We are trying to do minimized changes on the way of using HF.
* A smaller version may can't reveal the training loss convergence sometime.

**Comment by yanboliang** [comment]
> Hopefully, we can use this script in the future to publish numbers like the way Nvidia does it - https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch/LanguageModeling/BERT#results

Agree on what this benchmark would evolve toward!

File: benchmarks/training_loss.py (hunk 4 above)
Comment (voznesenskym) [reply benchmarks/training_loss.py:16]: Yeah, that seems reasonable, albeit unfortunate. 

cc @malfet for CI caching ideas?

File: benchmarks/training_loss.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    for batch in eval_dataloader:
+        batch = {k: v.to(device) for k, v in batch.items()}
+        with torch.no_grad():
+            outputs = opt_model(**batch)
+
+        logits = outputs.logits
+        predictions = torch.argmax(logits, dim=-1)
+        metric.add_batch(predictions=predictions, references=batch["labels"])
+
+    print("Loss history : " + str(loss_history))
```
Comment (anijain2305) [inline benchmarks/training_loss.py:89]: Can we add the code to tell if a backend PASSED or FAILED by storing the loss history of pytorch-native and the backend? Basically, this test will be super useful, if you just run from commandline, and it says PASS or FAIL. Otherwise, its too much eyeballing.

**Review by anijain2305 — CHANGES_REQUESTED** [review]

**Review by anijain2305 — APPROVED** [review]

File: benchmarks/training_loss.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    else:
+        print(
+            "[FAILED] TorchDynamo end to end training loss is greater than native Pytorch"
+        )
+    native_elapsed = native_end - native_start
+    dynamo_elapsed = dynamo_end - native_end
+    print(f"Train model on {args.epochs} epochs:")
+    print(f"PyTorch spent {timedelta(seconds=native_elapsed)}")
+    print(f"TorchDynamo spent {timedelta(seconds=dynamo_elapsed)}")
+
```
[the remark this answers is not in the document]
Comment (yanboliang) [reply benchmarks/training_loss.py:174]: I'll address this together with the following PRs.
